lib/indexer/indexer.js

```diff
@@ -0,0 +1,690 @@
+/*!
+ * indexer.js - abstract interface for bcoin indexers
+ * Copyright (c) 2018, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('assert');
+const path = require('path');
+const fs = require('bfile');
+const bio = require('bufio');
+const EventEmitter = require('events');
+const Logger = require('blgr');
+const Network = require('../protocol/network');
+const util = require('../utils/util');
+const layout = require('./layout');
+const CoinView = require('../coins/coinview');
+const Block = require('../primitives/block');
+const {ZERO_HASH} = require('../protocol/consensus');
+
+/**
+ * Indexer
+ * The class which indexers inherit from and implement the
+ * `indexBlock` and `unindexBlock` methods and database
+ * and storage initialization for indexing blocks.
+ * @alias module:indexer.Indexer
+ * @extends EventEmitter
+ * @abstract
+ */
+
+class Indexer extends EventEmitter {
+  /**
+   * Create an indexer.
+   * @constructor
+   * @param {String} module
+   * @param {Object} options
+   */
+
+  constructor(module, options) {
+    super();
+
+    assert(typeof module === 'string');
+    assert(module.length > 0);
+
+    this.options = new IndexOptions(module, options);
+
+    this.network = this.options.network;
+    this.logger = this.options.logger.context(`${module}indexer`);
+    this.blocks = this.options.blocks;
+    this.chain = this.options.chain;
+
+    this.db = null;
+    this.batch = null;
+    this.syncing = false;
+    this.height = 0;
+  }
+
+  /**
+   * Start a new batch write.
+   * @returns {Batch}
+   */
+
+  start() {
+    assert(this.batch === null, 'Already started.');
+    this.batch = this.db.batch();
+    return this.batch;
+  }
+
+  /**
+   * Put key and value to the current batch.
+   * @param {String} key
+   * @param {Buffer} value
+   */
+
+  put(key, value) {
+    this.batch.put(key, value);
+  }
+
+  /**
+   * Delete key from the current batch.
+   * @param {String} key
+   */
+
+  del(key) {
+    this.batch.del(key);
+  }
+
+  /**
+   * Commit the current batch.
+   * @returns {Promise}
+   */
+
+  async commit() {
+    await this.batch.write();
+    this.batch = null;
+  }
+
+  /**
+   * Open the indexer, open the database,
+   * initialize height, and bind to events.
+   * @returns {Promise}
+   */
+
+  async open() {
+    this.logger.info('Indexer is loading.');
+
+    await this.ensure();
+    await this.db.open();
+    await this.db.verify(layout.V.encode(), 'index', 0);
```

Comment:
If we have some way to represent indexer (unique identifier of some sort for the indexers, e.g. indexer name in db) we could also write that information to the db so indexer can double check the database belongs to it.
This is in case somehow txindexer (or other indexer) tries to open address indexer.
Otherwise indexers wont notice if they started to write to different indexer database.

Basically a way for Actual indexers to verify that the db belongs to it to some degree.
Where verifyNetwork makes sure you are not confusing networks
verifyIndexer would try to catch basic errors that could happen, this wont help if two indexers have same name but can cover for other errors.

Reply:
Hmmm... that isn't very likely, however it could probably be done using `this.module` as it's a string unique to the indexer and currently used to have the logs be unique.

Reply:
probably magic number is better for this purpose, but we will have to clarify how to come up with new ones so they don't copy/paste magic numbers.

Until we have better plugin system or separate indexers (that can run outside of bcoin), it wont be a problem.
Hacking into bcoin is not something we should encourage (unless "you know what you are doing.." ))

Reply:
Considering the `module` the unique identify would also derive the directory name, using it wouldn't avoid the conflict as it would be the same.

Reply:
There would be a _already_ open error also. This may not be a problem at this point.